lib/3scale/backend/stats/keys/generator/timestamp.rb

```diff
@@ -0,0 +1,126 @@
+            #     granularity being processed, and the granularity specified in the to limit does not
+            #     we iterate from the timestamp specified in the from limit, to the timestamp specified
+            #     in the service_context
+            # 4 - If there are limits, if the granularity specified in the from limit does not match the
+            #     granularity being processed, and the granularity specified in the to limit matches
+            #     we iterate from the timestamp specified in the service_context, to the timestamp specified
+            #     in the to limit
+            # 5 - If there are limits, if neither of the granularities specified in the from and to limits match
+            #     the granularity being processed, we iterate from the timestamp
+            #     to the timestamp specified in the service_context
+            def get_timestamp_limits(granularity_idx)
+              return service_context.from, service_context.to if index_limits.nil?
+
+              grn_start_lim, grn_end_lim = get_granularity_limits
+              ts_idx_lim = get_time_limits
+
+              from = service_context.from
+              to = service_context.to
+
+              # Only use limit from partition limits when partition index generator is current generator
+              from = ts_idx_lim.first if grn_start_lim == granularity_idx
+              to = ts_idx_lim.last if grn_end_lim == granularity_idx
+              [from, to]
+            end
+
+            def generate_timestamp_key(period)
+              granularity = period.granularity
+              key = "#{granularity}"
+              if granularity.to_sym != :eternity
+                key += ":#{period.start.to_compact_s}"
+              end
+              key
+            end
+
+            def get_datetime_generator_for_granularity(granularity_name, granularity_index)
+              Enumerator.new do |yielder|
+                from, to = get_timestamp_limits(granularity_index).map do |ts|
+                  ## A problem has been detected here in this two lines. It seems it is caching
+                  ## the @timestamp instance attribute for the Period[:month] even though
+                  ## we recreate the period. It caches the entire Period instance
+                  ## based on the start time, so because in this case both period
+                  ## objects have the same start time, it retrieves
+                  ## the object from the period cache with the same timestamp.
+                  ## Also, it is not possible to modify the timestamp of an existing
+                  ## period because it is a private field.
+                  start_ts = ThreeScale::Backend::Period[granularity_name].new(Time.at(ts).utc).start
+                  period = ThreeScale::Backend::Period[granularity_name].new(start_ts)
+                end
+                while from.start <= to.start
```

Comment:
while condition based on ts, hence no need to create Period instances for each step

Reply:
The declaration of the Period instances is needed in order to be able to use the 'start' and 'finish' methods available in the ThreeScale::Backend::Period class, which allows us to perform the increments in granularity steps